### ISSUE TITLE
Bagging very large data is faster with weights

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#885]
@guolinke Do you know why bagging using zero weights is faster than the original bagging in LightGBM?

Training one iteration is significantly faster with zero weights than using the bagging parameter.

May be related to this: https://github.com/Microsoft/LightGBM/issues/761

**Comment by guolinke** [comment]
@Laurae2 
what is bagging fraction you are using ?

**Comment by Laurae2** [comment]
@guolinke I'm using a 400GB (dense) dataset with 0.9 for `bagging_fraction` with 1 iteration only, repeated many times (using gradients only). Maybe the cost of initializing bagging is expensive? (my scenario is a very rare scenario anyway)

For 50GB datasets I always get better speed with bagging (any value).

I get a 1.05x speedup using weights arbitrary weights.

**Comment by guolinke** [comment]
@Laurae2 

The additional cost of bagging:
1. sampling in-bag indices and out-of-bag indices 
2. worse cache-missing problem when building histograms
3. additional cost for the prediction of out-of-bag data


So it is possible slower when #data is large and tree is small.

**Comment by Laurae2** [comment]
@guolinke thanks, now I get why bagging is slow in my case. As I do only 1 iteration and repeat this 500 times, it has the extra overhead everytime LightGBM runs (500 times).
